### ISSUE TITLE
Connection error on startup and GA4GH -- not proxy related

### ISSUE THREAD

**Issue by jblachly** [igvteam/igv#568]
_tl;dr: Two almost identical setups; one works great, while other produces connection error on startup and on attempt to access GA4GH. Not proxy related. Only difference in the two setups is system packages and libraries installed._

Hi, I have two computers, both with Ubuntu 16.4 LTS, both with OpenJDK 1.8.0_171, both with IGV (latest; 2.4.13, but problem occurs as early as 2.4.2 and possibly earlier). Both have identical bash environments (including correct http_ and https_ proxy settings), and identical IGV settings.

Interestingly, one starts up just fine, while the other throws a connection error (in 2.4.2 it showed two connection errors, one of which was related to GA4GH), and an error when trying to access GA4GH data.

Startup:
```
$ /tools/IGV_2.4.13/igv.sh
WARN [2018-07-20T16:41:18,828]  [Globals.java:138] [AWT-EventQueue-0]  Development mode is enabled
INFO [2018-07-20T16:41:18,864]  [DirectoryManager.java:179] [AWT-EventQueue-0]  IGV Directory: /home/OSUMC.EDU/blac96/igv
INFO [2018-07-20T16:41:18,898]  [Main.java:155] [AWT-EventQueue-0]  Startup  IGV Version 2.4.13 07/11/2018 02:16 PM
INFO [2018-07-20T16:41:18,899]  [Main.java:156] [AWT-EventQueue-0]  Java 1.8.0_171
INFO [2018-07-20T16:41:18,900]  [DirectoryManager.java:84] [AWT-EventQueue-0]  Fetching user directory...
INFO [2018-07-20T16:41:18,957]  [Main.java:157] [AWT-EventQueue-0]  Default User Directory: /home/OSUMC.EDU/blac96
INFO [2018-07-20T16:41:18,957]  [Main.java:158] [AWT-EventQueue-0]  OS: Linux


ERROR [2018-07-20T16:41:19,177]  [Main.java:212] [Thread-4]  Error checking version
java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:1.8.0_171]
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[?:1.8.0_171]
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[?:1.8.0_171]
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[?:1.8.0_171]
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:1.8.0_171]
        at java.net.Socket.connect(Socket.java:589) ~[?:1.8.0_171]
        at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:673) ~[?:1.8.0_171]
        at sun.net.NetworkClient.doConnect(NetworkClient.java:175) ~[?:1.8.0_171]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:463) ~[?:1.8.0_171]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:558) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:264) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:367) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1564) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492) ~[?:1.8.0_171]
        at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:347) ~[?:1.8.0_171]
        at org.broad.igv.util.HttpUtils.openConnection(HttpUtils.java:802) ~[igv.jar:?]
        at org.broad.igv.util.HttpUtils.openConnection(HttpUtils.java:675) ~[igv.jar:?]
        at org.broad.igv.util.HttpUtils.openConnection(HttpUtils.java:671) ~[igv.jar:?]
        at org.broad.igv.util.HttpUtils.getContentsAsString(HttpUtils.java:189) ~[igv.jar:?]
        at org.broad.igv.util.HttpUtils.getContentsAsString(HttpUtils.java:181) ~[igv.jar:?]
        at org.broad.igv.ui.Main.lambda$checkVersion$0(Main.java:190) ~[igv.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_171]
Jul 20, 2018 4:42:06 PM java.util.prefs.FileSystemPreferences$6 run
WARNING: Prefs file removed in background /home/OSUMC.EDU/blac96/.java/.userPrefs/prefs.xml
INFO [2018-07-20T16:42:08,466]  [GenomeManager.java:186] [pool-3-thread-1]  Loading genome: /home/OSUMC.EDU/blac96/igv/genomes/hg19.genome
INFO [2018-07-20T16:42:12,496]  [GenomeComboBox.java:79] [AWT-EventQueue-0]  Enter genome combo box
INFO [2018-07-20T16:42:12,507]  [GenomeManager.java:277] [pool-3-thread-1]  Genome loaded.  id= hg19
```

Because IGV version 2.4.2 also showed a startup connection error with a reference to GA4GH in the middle of the call stack, that led me to try `File>Load from Ga4gh`. On machine 1, it works fine, and the dialog is displayed. On the machine that complains at startup, it throws an additional connection error, and does not display the dialog:

```
ERROR [2018-07-20T16:42:16,429]  [BrowseGa4ghAction.java:98] [pool-3-thread-2]  Error opening Ga4gh dialog
java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:1.8.0_171]
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[?:1.8.0_171]
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[?:1.8.0_171]
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[?:1.8.0_171]
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:1.8.0_171]
        at java.net.Socket.connect(Socket.java:589) ~[?:1.8.0_171]
        at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:673) ~[?:1.8.0_171]
        at sun.security.ssl.BaseSSLSocketImpl.connect(BaseSSLSocketImpl.java:173) ~[?:1.8.0_171]
        at sun.net.NetworkClient.doConnect(NetworkClient.java:180) ~[?:1.8.0_171]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:463) ~[?:1.8.0_171]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:558) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:264) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:367) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1564) ~[?:1.8.0_171]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492) ~[?:1.8.0_171]
        at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480) ~[?:1.8.0_171]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:347) ~[?:1.8.0_171]
        at org.broad.igv.ga4gh.Ga4ghAPIHelper.handleHttpException(Ga4ghAPIHelper.java:301) ~[igv.jar:?]
        at org.broad.igv.ga4gh.Ga4ghAPIHelper.doPost(Ga4ghAPIHelper.java:293) ~[igv.jar:?]
        at org.broad.igv.ga4gh.Ga4ghAPIHelper.searchReadGroupsets(Ga4ghAPIHelper.java:105) ~[igv.jar:?]
        at org.broad.igv.ui.action.BrowseGa4ghAction$1.run(BrowseGa4ghAction.java:67) [igv.jar:?]
        at org.broad.igv.util.LongRunningTask.call(LongRunningTask.java:72) [igv.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_171]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_171]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_171]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_171]
```

I can confirm that the proxy settings are set correctly, and both machines can download new genomes, gene annotations, etc. just fine.

Although this is most likely related to different set of libraries on the two machines rather than a bug in IGV per se, I hope that one of the developers may recognize how the connection to GA4GH differs from other network connections being made so that I can fix my IGV installation.

**Comment by jrobinso** [comment]
Hi, its hard to say, but one difference is the GA4GH connections are by https  (ssl).   Many of the other connections are http.    I am inferring from your post and stack trace that IGV is able to make some connections (e.g. download the genome) but not others.  So my first guess would be SSL libraries.

If you do find this please update this thread.   It could be a rarely encountered IGV bug or incompatibility with some SSL libraries or settings.

**Comment by jblachly** [comment]
I will update this ticket as I learn more. I am fairly convinced this is a sort of overlap between bug and user-configuration.  That is to say, HTTPS/TLS works fine, _except when called by checkVersion() or from the GA4GH code_.

i.e., I am able to load tracks from amazon S3 over HTTPS, as well as other non-S3 providers (for example, ENCODE data portal).

The connection failures happen only when called from `org.broad.igv.ui.Main`
https://github.com/igvteam/igv/blob/master/src/main/java/org/broad/igv/ui/Main.java#L183

or from GA4GH (`org.broad.igv.ga4gh.Ga4ghAPIHelper.doPost`).

This suggests that the codepath there may be somehow different?

**Comment by jblachly** [comment]
SSL/TLS Package lists (excluding some perl packages) are identical:

`apt list --installed |grep -e ssl -e tls -e cert |grep -v perl`

```
ca-certificates/xenial-updates,xenial-updates,xenial-security,xenial-security,now 20170717~16.04.1 all [installed]
ca-certificates-java/xenial-updates,xenial-updates,now 20160321ubuntu1 all [installed,automatic]
libcurl3-gnutls/xenial-updates,xenial-security,now 7.47.0-1ubuntu2.8 amd64 [installed,automatic]
libcurl4-openssl-dev/xenial-updates,xenial-security,now 7.47.0-1ubuntu2.8 amd64 [installed]
libgnutls-dev/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed,automatic]
libgnutls-openssl27/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutls28-dev/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutls30/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutlsxx28/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed,automatic]
libssl-dev/xenial-updates,xenial-security,now 1.0.2g-1ubuntu4.13 amd64 [installed]
libssl-doc/xenial-updates,xenial-updates,xenial-security,xenial-security,now 1.0.2g-1ubuntu4.13 all [installed,automatic]
libssl1.0.0/xenial-updates,xenial-security,now 1.0.2g-1ubuntu4.13 amd64 [installed]
openssl/xenial-updates,xenial-security,now 1.0.2g-1ubuntu4.13 amd64 [installed]
openssl-blacklist/xenial,xenial,now 0.5-3 all [installed,automatic]
```

**Comment by jblachly** [comment]
Progress -- could this be a logical error in the check for proxy code?

I found the hidden setting DEBUG.PROXY. By enabling it, the machine that complains at startup / has slow startup, emits the following:

```
$ /tools/IGV_2.4.13/igv.sh
WARN [2018-07-25T11:25:35,730]  [Globals.java:138] [AWT-EventQueue-0]  Development mode is enabled
...
INFO [2018-07-25T11:25:35,842]  [Main.java:158] [AWT-EventQueue-0]  OS: Linux


INFO [2018-07-25T11:25:36,015]  [HttpUtils.java:468] [Thread-4]  Getting system proxy for https://data.broadinstitute.org/igv/projects/current/version.txt
INFO [2018-07-25T11:25:36,025]  [HttpUtils.java:756] [Thread-4]  PROXY NOT USED
ERROR [2018-07-25T11:25:36,026]  [Main.java:212] [Thread-4]  Error checking version
java.lang.NullPointerException: null

```

which means the following code path is triggered:
https://github.com/igvteam/igv/blob/2c98b738b4f1732c745245adfe274e9f19ffbbc5/src/main/java/org/broad/igv/util/HttpUtils.java#L754-L763

When retrieving other data (except GA4GH...) over HTTP or HTTPS, the proxy is used correctly.
```
INFO [2018-07-25T11:25:58,763]  [TrackLoader.java:132] [pool-3-thread-2]  Loading resource, path https://s3.amazonaws.com/igv.broadinstitute.org/annotations/hg19/genes/gencode.v18.annotation.sorted.gtf.gz
INFO [2018-07-25T11:25:58,771]  [HttpUtils.java:740] [pool-3-thread-2]  PROXY proxy.osumc.edu  8080
...
INFO [2018-07-25T11:25:59,228]  [HttpUtils.java:930] [pool-3-thread-2]  Range-byte request succeeded
```

**Comment by jrobinso** [comment]
Thanks for the thorough investigation!    Could you zip and attach the "prefs.properties" files for the machine that is working, and the machine that is not?   

Have you manually set the proxy settings in the IGV preferences window,  or are they picked up from the host system?   IGV trys to use the host proxy settings using the Java property  "java.net.useSystemProxies",  but this fails on occasion for unknown reasons.   That is why we have the option to set them manually.   If it does fail you might find the string  "Couldn't set useSystemProxies=true" in the igv.log,  but that's not guaranteed.  

There are a few reasons that the branch of code you reference could be run.   In the snippet below "igvProxySettingsExist" refer to the IGV preference settings.    

```java
        boolean igvProxySettingsExist = proxySettings != null && proxySettings.useProxy;
        boolean checkSystemProxy =
                !PreferencesManager.getPreferences().getAsBoolean("PROXY.DISABLE_CHECK") && !igvProxySettingsExist;

        //Only check for system proxy if igv proxy settings not found
        if (checkSystemProxy) {
            sysProxy = getSystemProxy(url.toExternalForm());
        }

        boolean useProxy =
                (sysProxy != null && sysProxy.type() != Proxy.Type.DIRECT) ||
                        (igvProxySettingsExist && !proxySettings.getWhitelist().contains(url.getHost()));
```
